core/src/packets/icmp/v6/ndp/mod.rs

```diff
@@ -19,105 +19,797 @@
 //! Neighbor Discovery Protocol
 //!
 //! NDP is a protocol used in IPv6, using ICMPv6 messages and operates at
-//! the link layer of the Internet model, as per [`IETF RFC 4861`]. It
+//! the link layer of the Internet model, as per [IETF RFC 4861]. It
 //! defines three mechanisms:
 //!
 //! - Substitute of ARP for use in IPv6 domains.
 //! - Stateless auto-configuration, allowing nodes on the local link to
 //!   configure their IPv6 addresses by themselves.
-//! - Router redirection to IPv6 nodes
+//! - Router redirection to IPv6 nodes.
 //!
-//! [`IETF RFC 4861`]: https://tools.ietf.org/html/rfc4861
+//! [IETF RFC 4861]: https://tools.ietf.org/html/rfc4861
 
 mod neighbor_advert;
 mod neighbor_solicit;
 mod options;
+mod redirect;
 mod router_advert;
 mod router_solicit;
 
 pub use self::neighbor_advert::*;
 pub use self::neighbor_solicit::*;
 pub use self::options::*;
+pub use self::redirect::*;
 pub use self::router_advert::*;
 pub use self::router_solicit::*;
 
-use crate::packets::Packet;
+use crate::dpdk::BufferError;
+use crate::packets::{Immutable, Internal, Packet};
+use crate::{ensure, Mbuf, SizeOf};
 use failure::Fallible;
+use std::fmt;
+use std::marker::PhantomData;
+use std::ptr::NonNull;
 
 /// A trait for common NDP accessors.
 pub trait NdpPacket: Packet {
     /// Returns the buffer offset where the options begin.
     fn options_offset(&self) -> usize;
 
-    /// Returns an iterator that iterates through the options in the NDP packet.
-    fn options(&self) -> NdpOptionsIterator<'_> {
-        let mbuf = self.mbuf();
+    /// Returns an iterator that iterates over the options in the packet.
+    ///
+    /// The options cannot be modified. To modify the values while iterating
+    /// over the options, use [`NdpOptions::iter`] instead.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let advert = ipv6.parse::<RouterAdvertisement<Ipv6>>()?;
+    /// let mut iter = advert.options();
+    ///
+    /// while let Some(option) = iter.next()? {
+    ///     println!("{:?}", option);
+    /// }
+    /// ```
+    ///
+    /// [`NdpOptions::iter`]: NdpOptions::iter
+    #[inline]
+    fn options(&self) -> ImmutableNdpOptionsIterator<'_> {
```

Comment:
@drunkirishcoder didn't we mention the move to `options_iter` here?

Reply:
yep and I forgot.